lib/chef/knife/ssh.rb

```diff
@@ -369,8 +375,14 @@ def ssh_command(command, subsession = nil)
                   ichannel.send_data("#{get_password}\n")
                 end
               end
+
+              ch.on_extended_data do |_, _type, data|
+                stderr += data
+              end
+
               ch.on_request "exit-status" do |ichannel, data|
                 exit_status = [exit_status, data.read_long].max
+                exit_status = open_session(subsession, command, true) if exit_status != 0 && stderr.include?("sudo: sorry, you must have a tty to run sudo")
```

Comment:
This should really raise, and the caller should rescue and then retry with pty=true

The problem here is that you're calling it recursively without exiting the subsession block so any ensure handlers aren't running and you're recursively calling open_channel with an already open_channel which is probably only going to find net-ssh bugs.

Reply:
Updated!